### ISSUE TITLE
Feature/Improvement request: model.fit(validation_split) with shuffling

### ISSUE THREAD

**Issue by mzoll** [keras-team/keras#10374]
I want to train my keras.model on a rather rich and complicated data set over some epochs, where I also want to make use of many of the nice features that are provided in Keras, like automated stopping.
However, I find that I really have problems to correctly address issue of overtraining through means of validation and early stopping in the currently provided `model.fit()` API. Namely the issue stems from the fact that an enabled `validation_split` (value >0.0) uses always the last n-percent samples from the input samples, explicitly _before_ shuffling (coimpare docu at https://keras.io/models/model/ ->fit ->validation_split ).

This means that even with `shuffle`-option enabled the validation will always be executed on the same (last) samples of the indata, so that one cannot perform cross-validation and early stopping based on the `val_loss` without it being skewed or biased towards these last samples.

I find it very limiting that the validation-split never can truly be a random selection, because this behavior can already be achieved by manually separating and providing a static validation sample by the parameter `validation_data` in the fit-function.

So my proposal would be to change the `fit()`-function so that it performs the separation of the validation-selection _after_ the the indata has been shuffled for each epoch, or to provide this functionality by an additional option.

**Comment by brge17** [comment]
Splitting data into train/validation/test is very application dependent.

The simplified UI ``validation_split``, is designed to give the user a deterministic result.

If you want to do things like cross validation and or other specific holdout rules, the best bet is to write your own data splitting outside ``Keras``.

**Comment by mzoll** [comment]
> The simplified UI validation_split, is designed to give the user a deterministic result.

This behavior, getting a deterministic result, can in the current UI already be achieved by specification of `validation_data`. This way is also much clearer, as the user is explicitly forced to make a static specification when he calls `fit()`. 

Secondly, the the UI  already contains the parameter option ´shuffle` with default `True`, which with the involved randomness already goes against your deterministic result intention (only valid in the limit of many repeated trainings or high statictics).

Writing a validation_split-off and reshuffling function outside the `fit()`-method to be executed before each training-epoch seems much more complicated, and will, as I see it, requires me to iteratively call to `fit(epochs=1)` in order to train for one epoch with reshuffled data once at a time. This renders the `callbacks` option, which's intent is to allow evaluations after each epoch, utterly useless, and would force me to also handle all that stuff from the outside, too.

On the other hand, making the proposed change, would allow the user to have some sort of in-build cross-validation in the `fit()`, or at least some measure for true result stability for the data, as training progresses over many epochs.

To sum up, changing the interface so that shuffle is applied _before_ `validation_split` would rather add a mode of operation (truely randomly shuffled train-data and validation-data) to the current fit-functionality, and make the prevalent mode (shuffled train-data, while static validation-data) more clear to the user by forcing its explicit specification by the already provided UI.

**Comment by briannemsick** [comment]
Shuffling the train/validation every Epoch is an ML mistake and is not cross validation. Think about it, the validation data for epoch2 will be in the training set for epoch1, meaning you will overestimate the model performance as compared to a true test set.

**Comment by mzoll** [comment]
Correct, it is not strictly a cross-validation, that's why I wrote "some sort of cross-validation". I guess that was still to strong a term for it, but it is still loosely comparable.
The thing is that currently the required strict separation is already kind of violated, when by the `validation_split`-option a certain fraction, and this being always the n-last unscrambled samples, is skimmed from the rest of the train_input_data. This makes the `shuffle`-option of `fit()` even more pointless, as for structured data I necessarily need to provide a shuffling from outside prior to me using the fit method, where one _again_ needs to specify `shuffle=True`.

Let me spin you a scenario to illustrate:
Lets say you have data, which stems from some sort of time-series. Most often such data would come with some progressive index, which kind of also helps to uniquely identify the samples, which is not used as a feature, but provides some internal structure to the data (think of desired operations of index-matching and row-merging and whatever).
Lets even say you engineer a wide variety of features for this data which are all stateless, so you have a rich dictionary of input tensors, something like `in_tensor_dict = { tensor0: np.array, tensor1: np.array, ... }` with arrays of shape nsamples, ndims.

Now, natively I would assume I could directly use this dictionary for training if I just remember that my NN should encounter shuffled data, so I would specify `mymodel.fit(in_tensor_dict, outcome, shuffle=True, epochs= e)`. Indeed this would work fine, but there is no way to control over-fitting on the data, then to stop the fit after each epoch and test.

Thus there should one wants to use the `validation_split`, so I can make use of validation metrices, like `val_loss`, which I can for example use for early stopping by a `callback[EarlyStopping(monitor='val_loss',...)]`. However, just _modifying_ my call to  `mymodel.fit(in_tensor_dict, outcome, shuffle=True, validation_split=0.1, callbacks=[EarlyStopping(monitor='val_loss')], epochs= sys.max_size)` would be wrong, as now my validation-data is effectively not represented in the effective train-data.

I find this very counter-intuitive as I explicitly specified `shuffle=True` and so assume shuffling. What I would have to do is to externally shuffle the samples in `in_tensor_dict` prior to calling `fit()`.

With the proposed change to shuffle _before_ splitting, the fit-algorithm would behave more _like_ a k-fold cross-validation, thus limiting the over-fitting on the train-data portion.
If it then is still my desire to test how well generalizing my model is, I can still have a held-out validation sample (maybe even from a temporal different time period) somewhere else and figure this out easily myself, by making one single evaluate/prediction call.

**Comment by brge17** [comment]
Exactly what you proposed, would lead to memorization of data for datasets such as``mnist`` or ``fashion_mnist``. Because the ``val_loss`` would constantly improve and the fake-validation data would become the training data for the next epoch.

``With the proposed change to shuffle before splitting, the fit-algorithm would behave more like an ``. That is untrue, after 2 epochs you have seen the entire data in training, **it behaves nothing like a k-fold cross validation**. It behaves more like having no ``validation_data`` at all.

**Comment by mzoll** [comment]
It is correct that with the NN learning from split-off val_data at some previous epoch, that the 'val_loss' would continuously improve; however, its absolute value is limited and the `val_loss` would approach, given enough training, some epsilon region.  The latter assumes that your model has not so many degrees of freedom, that it can ever learn to distinguish each data-point individually.

You could see it like this: The NN would be forced to learn the common generalizing features which are common to the randomly sampled train-data and the randomly sampled val-data. If it starts to learn features/characteristics, which are because of chance sampling only represented in the train-data but not in the val-data, the val-loss would get worse. When this happens you should, given enough patience, stop the training at that point.

It is also furthermore correct that after 2 epochs the training would have seen some, if not all, of the current or previous used validation sample. But as long you are aware of this being the case for this mode of operation and you still have a truly held out validation/test sample (which you actually should never hold attached to the train-data in the first place, as it this easily leads to mistakes), this will never be a problem.
If you desire to validate the model in training against such static held-out validation data, you can always just use the `validation_data` option, which makes the intention and purpose much clearer as it forces you specify it explicitly, e.g.
```python
val_frac = 0.1
split_idx=int(len(alldata)*val_frac)
model.fit(alldata[:sidx], alloutcome[:sidx], shuffle=True, validation_data=(alldata[sidx:], alloutcome[sidx:]) )
```
